### PR TITLE
[Runtime] Split server-client `run` skeleton

### PR DESCRIPTION
**Pull request by alonmr**
This is the start of the `run` separation between client and server.
This depicts the skeleton which will be implemented in the upcoming PRs.
Note PRs for this feature needs to be against branch `feature/run-separation`.
https://jira.iguazeng.com/browse/ML-2806

### CODE REVIEW

File: mlrun/api/launcher/server.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from mlrun.launcher.abstract import AbstractLauncher
+
+
+class ServerSideLauncher(AbstractLauncher):
```
Comment (liranbg) [inline mlrun/api/launcher/server.py:17]:
```suggestion
class Launcher(AbstractLauncher):
```

It is enough here as enyway it "lives" under "api" which essentially means it is a server-side launcher.

File: mlrun/api/launcher/server.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
```
Comment (liranbg) [inline mlrun/api/launcher/server.py:19]: not needed

File: mlrun/api/launcher/server.py (hunk 1 above)
Comment (liranbg) [reply mlrun/api/launcher/server.py:17]: also, I think you can move it alltogether under `mlrun/api/launcher.py`

File: mlrun/launcher/abstract.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    @staticmethod
+    def run(runtime):
+        """run the function from the server/client[local/remote]"""
+        pass
+
+    @staticmethod
+    @abc.abstractmethod
+    def _enrich_and_validate(runtime):
+        pass
```
Comment (liranbg) [inline mlrun/launcher/abstract.py:45]: separate them

File: mlrun/launcher/abstract.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    This class is designed to encapsulate the logic of running a function in different contexts
+    i.e. running a function locally, remotely or in a server
+    Each context will have its own implementation of the abstract methods
+    """
+
+    @staticmethod
+    @abc.abstractmethod
+    def verify_base_image(runtime):
+        """resolves whether build is required and sets the runtime base image"""
+        pass
```
Comment (liranbg) [inline mlrun/launcher/abstract.py:29]: `build_required`

File: mlrun/launcher/client/local.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from mlrun.launcher.abstract import AbstractLauncher
+
+
+class ClientLocalLauncher(AbstractLauncher):
```
Comment (liranbg) [inline mlrun/launcher/client/local.py:17]: Id expect the `mlrun/launcher` source tree to be like:

mlrun/launcher/factory.py
mlrun/aluncher/base.py
mlrun/aluncher/local.py
mlrun/aluncher/remote.py

class, w.r.t class names

base.py
`_BaseLauncher()`

local.py
`LocalLauncher()`

Comment (liranbg) [reply mlrun/launcher/client/local.py:17]: by that, you will remove many files/dirs that are overwhelming for now. also, keep the functions stubs to minimal, let them grow upon need / tasks

**Review by liranbg — CHANGES_REQUESTED** [review]

File: mlrun/launcher/factory.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import mlrun.errors
+from mlrun.launcher import ClientLocalLauncher, ClientRemoteLauncher, _BaseLauncher
+
+
+class LauncherFactory(object):
+    @staticmethod
+    def create_launcher(local) -> _BaseLauncher:
```
Comment (liranbg) [inline mlrun/launcher/factory.py:20]: type hint on local?

**Review by liranbg — APPROVED** [review]
very minor. lgtm! 🏅

**Review by Tankilevitch — COMMENTED** [review]
General wonder, seems a bit weird that common code will import code from the API, although the api launcher is backend side maybe it should be in the common code.
@liranbg @quaark @theSaarco wdyt?

**Comment by quaark** [comment]
> General wonder, seems a bit weird that common code will import code from the API, although the api launcher is backend side maybe it should be in the common code.
> @liranbg @quaark @theSaarco wdyt?

That's what we spoke about in the meeting. Yes this is a bit strange, but it's ok and we'll still be able to package the client and server separately as the import happens within the `if` statement.
I don't think `ServerSideLauncher` should sit in common, as that is exactly the server side implementation which should sit with the rest of the server code.

**Review by quaark — APPROVED** [review]

**Pull request merged**